penrose-web/src/compiler/Style.ts

```diff
@@ -1,7 +1,2068 @@
+  if (line.tag === "PathAssign") {
+    return {
+      ...line,
+      path: substitutePath(lv, subst, line.path),
+      value: substituteBlockExpr(lv, subst, line.value)
+    };
+  } else if (line.tag === "Override") {
+    return {
+      ...line,
+      path: substitutePath(lv, subst, line.path),
+      value: substituteBlockExpr(lv, subst, line.value)
+    };
+  } else if (line.tag === "Delete") {
+    return {
+      ...line,
+      contents: substitutePath(lv, subst, line.contents)
+    };
+  } else throw Error("Case should not be reached (anonymous statement should be substituted for a local one in `nameAnonStatements`)");
+};
+
+// Assumes a full substitution
+const substituteBlock = ([subst, si]: [Subst, number], [block, bi]: [Block, number], name: MaybeVal<string>): Block => {
+
+  const lvSubst: LocalVarSubst = name.tag === "Nothing"
+    ? { tag: "LocalVarId", contents: [bi, si] }
+    : { tag: "NamespaceId", contents: name.contents };
+
+  return {
+    ...block,
+    statements: block.statements.map(line => substituteLine(lvSubst, subst, line))
+  };
+};
+
+//#endregion Applying a substitution to a block
+
+// Convert Style expression to Substance expression (for ease of comparison in matching)
+const toSubExpr = (e: SelExpr): SubExpr => {
+  if (e.tag === "SEBind") {
+    return e.contents.contents;
+  } else if (e.tag === "SEFunc") {
+    return {
+      ...e, // Puts the remnants of e's ASTNode info here -- is that ok?
+      tag: "ApplyFunction",
+      name: e.name,
+      args: e.args.map(toSubExpr)
+    };
+  } else if (e.tag === "SEValCons") {
+    return {
+      ...e,
+      tag: "ApplyConstructor",
+      name: e.name,
+      args: e.args.map(toSubExpr)
+    };
+  } else if (e.tag === "SEFuncOrValCons") {
+    throw Error("compiler should have disambiguated func/val cons");
+  } else throw Error("unknown tag");
+};
+
+const toSubPredArg = (a: PredArg): SubPredArg => {
+  if (a.tag === "SEBind") {
+    return a.contents.contents;
+  } else if (a.tag === "RelPred") {
+    return toSubPred(a);
+  } else throw Error("unknown tag");
+};
+
+// Convert Style predicate to Substance predicate (for ease of comparison in matching)
+const toSubPred = (p: RelPred): ApplyPredicate => {
+  return {
+    ...p,
+    tag: "ApplyPredicate",
+    name: p.name,
+    args: p.args.map(toSubPredArg),
+  };
+};
+
+const varsEq = (v1: Identifier, v2: Identifier): boolean => {
+  return v1.value === v2.value;
+};
+
+const subVarsEq = (v1: Identifier, v2: Identifier): boolean => {
+  return v1.value === v2.value;
+};
+
+const argsEq = (a1: SubPredArg, a2: SubPredArg): boolean => {
+  if (a1.tag === "ApplyPredicate" && a2.tag === "ApplyPredicate") {
+    return subFnsEq(a1, a2);
+  } else if (a1.tag === a2.tag) { // both are SubExpr, which are not explicitly tagged
+    return subExprsEq(a1 as SubExpr, a2 as SubExpr);
+  } else return false; // they are different types
+};
+
+const subFnsEq = (p1: any, p2: any): boolean => {
```

Comment:
Why `any`? At least you'd know it's definitely an `ASTNode` (or something more specific)?